src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs

```diff
@@ -51,70 +50,171 @@ internal sealed class ServerMonitor : IServerMonitor
+                    cancellationTokenSource.Token),
+                cancellationTokenSource)
+        {
+        }
+
+        public ServerMonitor(ServerId serverId, EndPoint endPoint, IConnectionFactory connectionFactory, ServerMonitorSettings serverMonitorSettings, IEventSubscriber eventSubscriber, IRoundTripTimeMonitor roundTripTimeMonitor, CancellationTokenSource cancellationTokenSource)
         {
+            _monitorCancellationTokenSource = cancellationTokenSource;
             _serverId = Ensure.IsNotNull(serverId, nameof(serverId));
             _endPoint = Ensure.IsNotNull(endPoint, nameof(endPoint));
             _connectionFactory = Ensure.IsNotNull(connectionFactory, nameof(connectionFactory));
             Ensure.IsNotNull(eventSubscriber, nameof(eventSubscriber));
+            _serverMonitorSettings = Ensure.IsNotNull(serverMonitorSettings, nameof(serverMonitorSettings));
+
+            _baseDescription = _currentDescription = new ServerDescription(_serverId, endPoint, reasonChanged: "InitialDescription", heartbeatInterval: serverMonitorSettings.HeartbeatInterval);
+            _roundTripTimeMonitor = Ensure.IsNotNull(roundTripTimeMonitor, nameof(roundTripTimeMonitor));
 
-            _baseDescription = _currentDescription = new ServerDescription(_serverId, endPoint, reasonChanged: "InitialDescription", heartbeatInterval: heartbeatInterval);
-            _heartbeatInterval = heartbeatInterval;
-            _timeout = timeout;
             _state = new InterlockedInt32(State.Initial);
             eventSubscriber.TryGetEventHandler(out _heartbeatStartedEventHandler);
             eventSubscriber.TryGetEventHandler(out _heartbeatSucceededEventHandler);
             eventSubscriber.TryGetEventHandler(out _heartbeatFailedEventHandler);
             eventSubscriber.TryGetEventHandler(out _sdamInformationEventHandler);
         }
 
-        /// <inheritdoc />
         public ServerDescription Description => Interlocked.CompareExchange(ref _currentDescription, null, null);
 
+        public object Lock => _lock;
+
+        // public methods
+        public void CancelCurrentCheck()
+        {
+            IConnection toDispose = null;
+            lock (_lock)
+            {
+                if (!Ensure.IsNotNull(_currentCheckCancellationTokenSource, nameof(_currentCheckCancellationTokenSource)).IsCancellationRequested)
+                {
+                    _currentCheckCancellationTokenSource.Cancel();
+                    _currentCheckCancellationTokenSource.Dispose();
+                    _currentCheckCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(_monitorCancellationTokenSource.Token);
+                    // the previous isMaster cancelation token is still cancelled
+
+                    toDispose = _connection;
+                    _connection = null;
+                }
+            }
+            toDispose.Dispose();
```

Comment:
I think it's better to dispose `_connection` in line 314 https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/138/files#diff-df782a28b00d0b8b0d0f4cceae42dfe0R314. But at that moment we won't have a reference on `_connection`. So, I did it here